In [100]:
# Libraries

import pandas as pd
import numpy as np

In [101]:
# We have 3 different csv file, let's read those file individually

book = pd.read_csv('../data/Books.csv')
users = pd.read_csv('../data/Users.csv')
ratings = pd.read_csv('../data/Ratings.csv')


/tmp/ipykernel_15080/688384193.py:3: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  book = pd.read_csv('../data/Books.csv')


In [102]:
book.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [103]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [104]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [105]:
# Let's find out the shape of all csv file

def shape(df):
    print(df.shape)

In [106]:
shape(book)

(271360, 8)


In [107]:
shape(users)

(278858, 3)


In [108]:
shape(ratings)

(1149780, 3)


In [109]:
# let's check null values using similar war

def isNull(dataframe):
    return dataframe.isnull().sum()

In [110]:
isNull(book)

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

We can see there are some null values in this csv file. We will tackle it later.

In [111]:
isNull(users)

User-ID          0
Location         0
Age         110762
dtype: int64

There is huge number of missing values in age column.

In [112]:
isNull(ratings)

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

We are good to go in this dataframe.


In [113]:
book.duplicated().sum()

np.int64(0)

In [114]:
book.nunique()

ISBN                   271360
Book-Title             242135
Book-Author            102022
Year-Of-Publication       202
Publisher               16807
Image-URL-S            271044
Image-URL-M            271044
Image-URL-L            271041
dtype: int64

# Popularity Based Recommender System

`Popularity based recommender system` : These are like trending videos in youtube. It stores bunch of popular things in one platform and is available for user. Some user may not like it.

In [115]:
# let's see books with highest rating
rating_with_name = ratings.merge(book, on='ISBN')
rating_with_name.head()

User-ID        ISBN  Book-Rating  \
0   276725  034545104X            0   
1   276726  0155061224            5   
2   276727  0446520802            0   
3   276729  052165615X            3   
4   276729  0521795028            6   

                                          Book-Title      Book-Author  \
0                               Flesh Tones: A Novel       M. J. Rose   
1                                   Rites of Passage       Judith Rae   
2                                       The Notebook  Nicholas Sparks   
3                                     Help!: Level 1    Philip Prowse   
4  The Amsterdam Connection : Level 4 (Cambridge ...      Sue Leather   

  Year-Of-Publication                   Publisher  \
0                2002            Ballantine Books   
1                2001                      Heinle   
2                1996                Warner Books   
3                1999  Cambridge University Press   
4                2001  Cambridge University Press   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/0155061224.0...   
2  http://images.amazon.com/images/P/0446520802.0...   
3  http://images.amazon.com/images/P/052165615X.0...   
4  http://images.amazon.com/images/P/0521795028.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/0155061224.0...   
2  http://images.amazon.com/images/P/0446520802.0...   
3  http://images.amazon.com/images/P/052165615X.0...   
4  http://images.amazon.com/images/P/0521795028.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/0155061224.0...  
2  http://images.amazon.com/images/P/0446520802.0...  
3  http://images.amazon.com/images/P/052165615X.0...  
4  http://images.amazon.com/images/P/0521795028.0...

In [116]:
num_ratingin_df = rating_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index() # reset_index returns the dataframe

In [117]:
num_ratingin_df.rename(columns={"Book-Rating": 'num_rating'}, inplace=True)

In [118]:
num_ratingin_df

Book-Title  num_rating
0        A Light in the Storm: The Civil War Diary of ...           4
1                                   Always Have Popsicles           1
2                    Apple Magic (The Collector's series)           1
3        Ask Lily (Young Women of Faith: Lily Series, ...           1
4        Beyond IBM: Leadership Marketing and Finance ...           1
...                                                   ...         ...
241066                                      Ã?Â?lpiraten.           2
241067                     Ã?Â?rger mit Produkt X. Roman.           4
241068                                Ã?Â?sterlich leben.           1
241069                              Ã?Â?stlich der Berge.           3
241070                                  Ã?Â?thique en toc           2

[241071 rows x 2 columns]

In [119]:
# Ensure 'Book-Rating' is numeric
rating_with_name['Book-Rating'] = pd.to_numeric(rating_with_name['Book-Rating'], errors='coerce')


In [120]:
# Group by 'Book-Title' and calculate the mean rating
avg_rating_df = rating_with_name.groupby('Book-Title')['Book-Rating'].agg(lambda x: x.astype(float).mean()).reset_index()

In [121]:
# Renaming Book-rating into avg_rating
avg_rating_df.rename(columns={'Book-Rating':'avg_rating'},inplace=True)
avg_rating_df

Book-Title  avg_rating
0        A Light in the Storm: The Civil War Diary of ...    2.250000
1                                   Always Have Popsicles    0.000000
2                    Apple Magic (The Collector's series)    0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...    8.000000
4        Beyond IBM: Leadership Marketing and Finance ...    0.000000
...                                                   ...         ...
241066                                      Ã?Â?lpiraten.    0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.    5.250000
241068                                Ã?Â?sterlich leben.    7.000000
241069                              Ã?Â?stlich der Berge.    2.666667
241070                                  Ã?Â?thique en toc    4.000000

[241071 rows x 2 columns]

In [122]:
popular_df = num_ratingin_df.merge(avg_rating_df, on='Book-Title')
popular_df



Book-Title  num_rating  \
0        A Light in the Storm: The Civil War Diary of ...           4   
1                                   Always Have Popsicles           1   
2                    Apple Magic (The Collector's series)           1   
3        Ask Lily (Young Women of Faith: Lily Series, ...           1   
4        Beyond IBM: Leadership Marketing and Finance ...           1   
...                                                   ...         ...   
241066                                      Ã?Â?lpiraten.           2   
241067                     Ã?Â?rger mit Produkt X. Roman.           4   
241068                                Ã?Â?sterlich leben.           1   
241069                              Ã?Â?stlich der Berge.           3   
241070                                  Ã?Â?thique en toc           2   

        avg_rating  
0         2.250000  
1         0.000000  
2         0.000000  
3         8.000000  
4         0.000000  
...            ...  
241066    0.000000  
241067    5.250000  
241068    7.000000  
241069    2.666667  
241070    4.000000  

[241071 rows x 3 columns]

In [123]:
# Books with num_rating > 250

popular_df = popular_df[popular_df['num_rating'] >= 250].sort_values('avg_rating', ascending=False).head(20)
popular_df

Book-Title  num_rating  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)         428   
80422        Harry Potter and the Goblet of Fire (Book 4)         387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)         278   
80426   Harry Potter and the Order of the Phoenix (Boo...         347   
80414    Harry Potter and the Chamber of Secrets (Book 2)         556   
191612  The Hobbit : The Enchanting Prelude to The Lor...         281   
187377  The Fellowship of the Ring (The Lord of the Ri...         368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...         575   
211384     The Two Towers (The Lord of the Rings, Part 2)         260   
219741                              To Kill a Mockingbird         510   
183573                                  The Da Vinci Code         898   
187880                 The Five People You Meet in Heaven         430   
180556                             The Catcher in the Rye         449   
196326                          The Lovely Bones: A Novel        1295   
764                                                  1984         284   
144165                           Prodigal Summer: A Novel         253   
128670                                         Neverwhere         265   
206502                            The Secret Life of Bees         774   
168719  Stupid White Men ...and Other Sorry Excuses fo...         283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...         493   

        avg_rating  
80434     5.852804  
80422     5.824289  
80441     5.737410  
80426     5.501441  
80414     5.183453  
191612    5.007117  
187377    4.948370  
80445     4.895652  
211384    4.880769  
219741    4.700000  
183573    4.642539  
187880    4.551163  
180556    4.545657  
196326    4.468726  
764       4.454225  
144165    4.450593  
128670    4.449057  
206502    4.447028  
168719    4.356890  
223135    4.354970

In [124]:
# drop duplicate book-title
popular_df = popular_df.merge(book, on='Book-Title').drop_duplicates('Book-Title')

In [125]:
popular_df

Book-Title  num_rating  avg_rating  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)         428    5.852804   
3        Harry Potter and the Goblet of Fire (Book 4)         387    5.824289   
5      Harry Potter and the Sorcerer's Stone (Book 1)         278    5.737410   
9   Harry Potter and the Order of the Phoenix (Boo...         347    5.501441   
13   Harry Potter and the Chamber of Secrets (Book 2)         556    5.183453   
16  The Hobbit : The Enchanting Prelude to The Lor...         281    5.007117   
17  The Fellowship of the Ring (The Lord of the Ri...         368    4.948370   
26  Harry Potter and the Sorcerer's Stone (Harry P...         575    4.895652   
28     The Two Towers (The Lord of the Rings, Part 2)         260    4.880769   
39                              To Kill a Mockingbird         510    4.700000   
47                                  The Da Vinci Code         898    4.642539   
53                 The Five People You Meet in Heaven         430    4.551163   
55                             The Catcher in the Rye         449    4.545657   
62                          The Lovely Bones: A Novel        1295    4.468726   
63                                               1984         284    4.454225   
72                           Prodigal Summer: A Novel         253    4.450593   
73                                         Neverwhere         265    4.449057   
78                            The Secret Life of Bees         774    4.447028   
84  Stupid White Men ...and Other Sorry Excuses fo...         283    4.356890   
85  Tuesdays with Morrie: An Old Man, a Young Man,...         493    4.354970   

          ISBN         Book-Author Year-Of-Publication  \
0   0439136350       J. K. Rowling                1999   
3   0439139597       J. K. Rowling                2000   
5   0590353403       J. K. Rowling                1998   
9   043935806X       J. K. Rowling                2003   
13  0439064872       J. K. Rowling                2000   
16  0345339681      J.R.R. TOLKIEN                1986   
17  0345339703      J.R.R. TOLKIEN                1986   
26  059035342X       J. K. Rowling                1999   
28  0345339711      J.R.R. TOLKIEN                1986   
39  0446310786          Harper Lee                1988   
47  0385504209           Dan Brown                2003   
53  0786868716         Mitch Albom                2003   
55  0316769487       J.D. Salinger                1991   
62  0316666343        Alice Sebold                2002   
63  0451524934       George Orwell                1990   
72  0060959037  Barbara Kingsolver                2001   
73  0380789019         Neil Gaiman                1998   
78  0142001740       Sue Monk Kidd                2003   
84  0060392452       Michael Moore                2002   
85  0385484518         MITCH ALBOM                1997   

                     Publisher  \
0                   Scholastic   
3                   Scholastic   
5                   Scholastic   
9                   Scholastic   
13                  Scholastic   
16                     Del Rey   
17                     Del Rey   
26      Arthur A. Levine Books   
28                     Del Rey   
39  Little Brown &amp; Company   
47                   Doubleday   
53                    Hyperion   
55               Little, Brown   
62               Little, Brown   
63                 Signet Book   
72                   Perennial   
73                        Avon   
78               Penguin Books   
84                 Regan Books   
85                   Doubleday   

                                          Image-URL-S  \
0   http://images.amazon.com/images/P/0439136350.0...   
3   http://images.amazon.com/images/P/0439139597.0...   
5   http://images.amazon.com/images/P/0590353403.0...   
9   http://images.amazon.com/images/P/043935806X.0...   
13  http://images.amazon.com/images/P/0439064872.0...   
16  http://images.amazon.com/images/P/0345339681.0...   
17  http

In [126]:
# Take only useful column from popular_df
popular_df = popular_df[['Book-Title', 'Book-Author', 'Image-URL-M', 'num_rating', 'avg_rating']]

In [127]:
popular_df.head()

Book-Title    Book-Author  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)  J. K. Rowling   
3        Harry Potter and the Goblet of Fire (Book 4)  J. K. Rowling   
5      Harry Potter and the Sorcerer's Stone (Book 1)  J. K. Rowling   
9   Harry Potter and the Order of the Phoenix (Boo...  J. K. Rowling   
13   Harry Potter and the Chamber of Secrets (Book 2)  J. K. Rowling   

                                          Image-URL-M  num_rating  avg_rating  
0   http://images.amazon.com/images/P/0439136350.0...         428    5.852804  
3   http://images.amazon.com/images/P/0439139597.0...         387    5.824289  
5   http://images.amazon.com/images/P/0590353403.0...         278    5.737410  
9   http://images.amazon.com/images/P/043935806X.0...         347    5.501441  
13  http://images.amazon.com/images/P/0439064872.0...         556    5.183453

# Collaborative Filtering Based Recommender System


In [132]:
# peoples and the number of books they rated

num_book_rated = rating_with_name.groupby('User-ID').count()['Book-Rating']
num_book_rated

User-ID
2          1
8         17
9          3
10         1
12         1
          ..
278846     1
278849     4
278851    23
278852     1
278854     8
Name: Book-Rating, Length: 92106, dtype: int64

user id 2 has rated 1 book, 8 has rated 17 books and so on.

In [133]:
# people only with more than 200 book rated
book_rated_200 = num_book_rated > 200

In [139]:
litterate_users = book_rated_200[book_rated_200].index
litterate_users

Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,   6323,
         6543,
       ...
       271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427, 277639,
       278418],
      dtype='int64', name='User-ID', length=811)

User_id of the users

In [142]:
# Litterate users
filtered_users = rating_with_name[rating_with_name['User-ID'].isin(litterate_users)]
filtered_users

User-ID        ISBN  Book-Rating  \
1150      277427  002542730X           10   
1151      277427  0026217457            0   
1152      277427  003008685X            8   
1153      277427  0030615321            0   
1154      277427  0060002050            0   
...          ...         ...          ...   
1029357   275970  1931868123            0   
1029358   275970  3411086211           10   
1029359   275970  3829021860            0   
1029360   275970  4770019572            0   
1029361   275970  9626340762            8   

                                                Book-Title  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
1151                    Vegetarian Times Complete Cookbook   
1152                                              Pioneers   
1153      Ask for May, Settle for June (A Doonesbury book)   
1154                     On a Wicked Dawn (Cynster Novels)   
...                                                    ...   
1029357  There's a Porcupine in My Outhouse: Misadventu...   
1029358                                         Die Biene.   
1029359                                     The Penis Book   
1029360                                            Musashi   
1029361  Northanger Abbey (Classic Literature with Clas...   

                           Book-Author Year-Of-Publication  \
1150                 James Finn Garner                1994   
1151                        Lucy  Moll                1995   
1152             James Fenimore Cooper                1974   
1153                     G. B. Trudeau                1982   
1154                 Stephanie Laurens                2002   
...                                ...                 ...   
1029357                   Mike Tougias                2002   
1029358  Sybil Gr�¤fin Sch�¶nfeldt                1993   
1029359                   Joseph Cohen                1999   
1029360                 Eiji Yoshikawa                1995   
1029361                    Jane Austen                1996   

                                    Publisher  \
1150                John Wiley &amp; Sons Inc   
1151                    John Wiley &amp; Sons   
1152                         Thomson Learning   
1153                      Henry Holt &amp; Co   
1154                               Avon Books   
...                                       ...   
1029357                    Capital Books (VA)   
1029358  Bibliographisches Institut, Mannheim   
1029359                              Konemann   
1029360          Kodansha International (JPN)   
1029361                 Naxos Audiobooks Ltd.   

                                               Image-URL-S  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1151     http://images.amazon.com/images/P/0026217457.0...   
1152     http://images.amazon.com/images/P/003008685X.0...   
1153     http://images.amazon.com/images/P/0030615321.0...   
1154     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1029357  http://images.amazon.com/images/P/1931868123.0...   
1029358  http://images.amazon.com/images/P/3411086211.0...   
1029359  http://images.amazon.com/images/P/3829021860.0...   
1029360  http://images.amazon.com/images/P/4770019572.0...   
1029361  http://images.amazon.com/images/P/9626340762.0...   

                                               Image-URL-M  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1151     http://images.amazon.com/images/P/0026217457.0...   
1152     http://images.amazon.com/images/P/003008685X.0...   
1153     http://images.amazon.com/images/P/0030615321.0...   
1154     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1029357  http://images.amazon.com/images/P/1931868123.0...   
1029358  http://images.amazon.com/images/P/3411086211.0...   
1029359  http://images.amazon.com/images/P/3829021860.0...   
1029360  http://images.a

Out of 1149780, only 474007 are litterate users meaning user with rated more than 200 books.

In [143]:
# book with 50 + rating

y = filtered_users.groupby('Book-Title').count()['Book-Rating']>=50
famous_book = y[y].index



In [144]:
famous_book

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

In [146]:
final_rating = filtered_users[filtered_users['Book-Title'].isin(famous_book)]

In [149]:
pt = final_rating.pivot_table(index='Book-Title', columns='User-ID', values='Book-Rating')

In [150]:
# Fill pivot table 
pt.fillna(0, inplace=True)

In [151]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

### Time to find out the distance from one vector to another vector, We need it to measures the similarity between two vectors of an inner product space.

In [152]:
from sklearn.metrics.pairwise import cosine_similarity



In [153]:
similarity_score = cosine_similarity(pt)

In [156]:
similarity_score.shape

(706, 706)

It is the equlian distance of 706 books with 706 books 

In [157]:
similarity_score[0]

array([1.        , 0.10255025, 0.01220856, 0.        , 0.05367224,
       0.02774901, 0.08216491, 0.13732869, 0.03261686, 0.03667591,
       0.02322418, 0.06766487, 0.02083978, 0.09673735, 0.13388865,
       0.08303112, 0.11153543, 0.05100411, 0.02517784, 0.11706383,
       0.        , 0.14333793, 0.07847534, 0.06150451, 0.08723968,
       0.        , 0.07009814, 0.13658681, 0.07600328, 0.12167134,
       0.00768046, 0.01473221, 0.        , 0.07965814, 0.04522617,
       0.01556271, 0.09495938, 0.0182307 , 0.02610465, 0.07984012,
       0.11679969, 0.0569124 , 0.08354155, 0.08471898, 0.08785938,
       0.05491435, 0.0548505 , 0.27026514, 0.09779123, 0.06016046,
       0.08958835, 0.06748675, 0.        , 0.04468098, 0.01920872,
       0.        , 0.05629067, 0.00557964, 0.07877059, 0.05219479,
       0.18908177, 0.        , 0.01240656, 0.02984572, 0.04279502,
       0.12680125, 0.16566735, 0.        , 0.13357242, 0.06615478,
       0.        , 0.        , 0.        , 0.10968075, 0.02806

Similarity score with index 0 and other remaining other. It returns 1 in first index because index 0 is exactly same when comparing with itself.

In [178]:
# Function to recommend book

def recommend(book_name):
    # fetch index
    index = np.where(pt.index==book_name)[0][0]
    # 5 books with simlar items
    similar_items = sorted(list(enumerate(similarity_score[index])),key=lambda x:x[1],reverse=True)[1:6]

    # let's print the books using loop
    for i in similar_items:
        print(pt.index[i[0]])



In [180]:
recommend("River's End")

Sanctuary
Midnight Bayou
Heart of the Sea (Irish Trilogy)
Jewels of the Sun (Irish Trilogy)
Macgregor Brides (Macgregors)


In [160]:
# fetch index
index = np.where(pt.index=='Zoya')[0][0]
index

np.int64(704)

In [166]:
# 5 books with simlar items
similar_items = sorted(list(enumerate(similarity_score[index])),key=lambda x:x[1],reverse=True)[1:5]